### PR TITLE
Extend ReturnOverrides

### PR DESCRIPTION
**Pull request by matiasinsaurralde**
See #762.

As a sample plugin (test_app.py)

```python
from tyk.decorators import *

@Hook
def MyCustomMiddleware(request, session, spec):
    print("my_middleware: MyCustomMiddleware")
    request.object.return_overrides.headers['content-type'] = 'application/json'
    request.object.return_overrides.response_code = 200
    request.object.return_overrides.response_error = "{\"key\": \"value\"}\n"
    return request, session
```

With manifest:
```json
{
  "file_list": [
    "test_app.py"
  ],
  "custom_middleware": {
    "post": [
      {"name": "MyCustomMiddleware"}
    ],
    "driver": "python"
  }
}
```

Curl output:
```
% curl localhost:8080/test-api/headers -v
*   Trying ::1...
* Connected to localhost (::1) port 8080 (#0)
> GET /test-api/headers HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.43.0
> Accept: */*
> 
< HTTP/1.1 200
< Content-Type: application/json
< Date: Wed, 24 May 2017 08:32:50 GMT
< Content-Length: 17
< 
{"key": "value"}
* Connection #0 to host localhost left intact
```

### CODE REVIEW

File: coprocess.go [hunk 1]
```diff
@@ -285,6 +285,15 @@ func (m *CoProcessMiddleware) ProcessRequest(w http.ResponseWriter, r *http.Requ
 		return errors.New("Key not authorised"), int(returnObject.Request.ReturnOverrides.ResponseCode)
 	}
 
+	if returnObject.Request.ReturnOverrides.ResponseCode > 0 {
+		for h, v := range returnObject.Request.ReturnOverrides.Headers {
+			w.Header().Set(h, v)
+		}
+		w.WriteHeader(int(returnObject.Request.ReturnOverrides.ResponseCode))
+		w.Write([]byte(returnObject.Request.ReturnOverrides.ResponseError))
+		return nil, 666
```
Comment (mvdan) [inline coprocess.go]: What does `666` mean here?

Comment (lonelycode) [inline coprocess.go]: 666 is a magic return code that makes the gateway halt MW processing and respond

Comment (mvdan) [inline coprocess.go]: Ah, I see we have `1666` also - we should use constants for these to make them more obvious, like `return nil, mwStatusRespond`. I'll ticket that up to do once this is merged.

Comment (lonelycode) [inline coprocess.go]: That's a fantastic idea, let's do that! So much clearer

**Comment by matiasinsaurralde** [comment]
@mvdan Just updated the tests, let's wait for #764 to replace the integers with constants.
I'm also extending the rest of the CP tests, trying to enhance them, will be a separate PR.

File: coprocess_test.go [hunk 2]
```diff
@@ -247,6 +247,33 @@ func TestCoProcessAuth(t *testing.T) {
...
+	spec := createSpecTest(t, basicCoProcessDef)
+
+	chain := buildCoProcessChain(spec, "hook_test_return_overrides", coprocess.HookType_Pre, apidef.MiddlewareDriver("python"))
+
+	session := createNonThrottledSession()
+	spec.SessionManager.UpdateSession("abc", session, 60)
+
+	recorder := httptest.NewRecorder()
+
+	req, err := http.NewRequest("GET", "/headers", strings.NewReader(""))
```
Comment (mvdan) [inline coprocess_test.go]: You can use a `nil` body

File: coprocess_test_helpers.go [hunk 3]
```diff
@@ -125,6 +125,7 @@ func TestTykTriggerEvent(eventName, eventPayload string) {
 
 //export applyTestHooks
 func applyTestHooks(objectPtr unsafe.Pointer) {
+
```
Comment (mvdan) [inline coprocess_test_helpers.go]: rogue empty line?

File: coprocess_test.go (hunk 2 above)
Comment (matiasinsaurralde) [inline coprocess_test.go]: A `nil` body panics 😮

Comment (mvdan) [inline coprocess_test.go]: Something is buggy somewhere... forget I said anything for now :)

**Comment by buger** [comment]
@matiasinsaurralde can we fix tests?

**Comment by mvdan** [comment]
@buger tests are fixed, FYI

**Comment by buger** [comment]
I'm right that this is similar to what we have now with Virtual Endpoints in our JS middleware?
Previously we already had option to add new headers, but with this functionality it is now possible to write any logic. Right?

Also, does it affect only Coprocess (Lua, Python) plugins, or gRPC ones too (because I see that you changed protobuffers files as well)?

**Comment by matiasinsaurralde** [comment]
I'm not very familiar with the virtual endpoint feature. Yes, basically it's possible to take control of the request flow and return a custom body.
There is an existing feature to add headers from the CP middleware, but it's for request headers (before they hit the upstream API). The funcionality covered by this PR is specific to the response object, and the headers that are specified here are received by the client.
It affects all the plugins and will work as long as you're using the latest definitions and/or bindings provided by us.

**Pull request merged**